### PR TITLE
KAFKA-6521: Store record timestamps in KTable stores

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mjsax** [comment]
Contains a lot of Java8 cleanup etc.

The actual store changes happens in `KeyValueStoreMaterializer.java`.

This implies, that all `Processors` using a store need to be updated from `KeyValueStore` to `KeyValueWithTimestampStore`. The PR include changes to leverage the timestamp to compute timestamp for the output records to get deterministic semantics.

We also need to update "KTable Getter", and forwarding tuple logic to carry forward the correct timestamp. The `CachingStore` needs a slight hack (this part might actually be required for the original PR for KIP-258 already). I hit a couple of other minor bugs, that need to be included there.

Test needed to be updated to expect `ValueAndTimestamp<V>` instead of `V` -- also, we need to set a timestamp when piping data throw the topology to get deterministic results.

The beauty of this PR is, that is after KIP-258 is merged, we can just redirect to AK/trunk to address KAFKA-6521 (it's touches almost only DSL code and should not be hard to rebase/clean up :) )

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java [hunk 1]
```diff
@@ -122,11 +124,11 @@ public void process(final K key, final V value) {
             if (!mergedWindow.equals(newSessionWindow)) {
                 for (final KeyValue<Windowed<K>, Agg> session : merged) {
                     store.remove(session.key);
-                    tupleForwarder.maybeForward(session.key, null, session.value);
+                    tupleForwarder.maybeForward(session.key, null, session.value, context().timestamp()); // TODO set proper result timestamp
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:127]: If not `context().timestamp()` where would we get the correct timestamp for comparison/calculation?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java [hunk 2]
```diff
@@ -67,21 +67,19 @@ public void init(final ProcessorContext context, final StateStore root) {
...
     }
 
-    @SuppressWarnings("unchecked")
+    //@SuppressWarnings("unchecked")
     private void initInternal(final ProcessorContext context) {
         this.context = (InternalProcessorContext) context;
-        this.serdes = new StateSerdes<>(ProcessorStateManager.storeChangelogTopic(context.applicationId(), underlying.name()),
-                                        keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
-                                        valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+        serdes = new StateSerdes<>(
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:73]: more of a meta-comment, but do we want to piggyback the work of not using the cache at all if the user specifies `cache max bytes = 0`? I'm guessing not, probably better in a separate PR, but I just wanted to note that.

**Review by bbejeck — COMMENTED** [review]
Thanks for doing this @mjsax.  I did a quick pass and this looks good! I left a couple of questions more out my own curiosity.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:127]: Atm, `SessionWithTimestampStore` is missing. After we added it, we can do the same as for other operators and use `max(storedTimestamp, recordTimestamp)`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java:73]: Yes, we should do this in a separate PR (seem unrelated). I think, it's a good idea for the optimization layer -- we pass in the config, and can check the setting. If set to zero, we can call `.disableCaching()` for all stores before building the topology. Can you create a Jira for this for tracking? Thx.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:127]: ah, yeah thanks for clearing up, kind of obvious in hindsight

**Pull request closed without merging**